nixos/modules/services/networking/lxd-image-server.nix

```diff
@@ -0,0 +1,142 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.lxd-image-server;
+
+  location = "/var/www/simplestreams";
+in
+{
+  options = {
+    services.lxd-image-server = {
+      enable = mkEnableOption "lxd-image-server";
+
+      group = mkOption {
+        type = types.str;
+        description = "Group assigned to the user and the webroot directory";
+        default = "nginx";
+        example = "www-data";
+      };
+
+      extraConfig = mkOption {
+        type = with types; lines;
+        description = ''
+          Configuration for lxd-image-server
+
+          Example see https://github.com/Avature/lxd-image-server/blob/master/config.toml
+        '';
+        default = "";
+      };
+
+      nginx = {
+        enable = mkEnableOption "nginx";
+        domain = mkOption {
+          type = types.str;
+          description = "Domain to use for nginx virtual host";
+          example = "images.example.org";
+        };
+      };
+    };
+  };
+
+  config = mkMerge [
+    (mkIf (cfg.enable) {
+      users.users.lxd-image-server = {
+        isSystemUser = true;
+        group = cfg.group;
+      };
+      users.groups.${cfg.group} = {};
+
+      environment.etc."lxd-image-server/config.toml".text = cfg.extraConfig;
+
+      services.logrotate.extraConfig = ''
+        /var/log/lxd-image-server/lxd-image-server.log {
+            daily
+            missingok
+            rotate 21
+            compress
+            delaycompress
+            copytruncate
+            notifempty
+        }
+      '';
+
+      systemd.services.lxd-image-server = {
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+
+        description = "LXD Image Server";
+
+        preStart = ''
+          LOC=${escapeShellArg location}
+          if [ ! -d "$LOC" ]; then
+            mkdir -p "$LOC"
+          fi
+          chown lxd-image-server:${cfg.group} "$LOC"
+          cd "$LOC"
+
+          chown -R lxd-image-server:${cfg.group} /var/log/lxd-image-server
+          chown -R lxd-image-server:${cfg.group} /var/www/simplestreams
+          touch /var/log/lxd-image-server/lxd-image-server.log
+          chown lxd-image-server:${cfg.group} /var/log/lxd-image-server/lxd-image-server.log
+          chown -R lxd-image-server:${cfg.group} /var/run/lxd-image-server
+
+          ${pkgs.lxd-image-server}/bin/lxd-image-server init
+        '';
+
+        serviceConfig = {
+          User = "lxd-image-server";
+          Group = cfg.group;
+          PermissionsStartOnly = true;
```

Comment:
This was deprecated quite a while ago. Please remove.

Reply:
is there any suitable replacement?

Reply:
I'm assuming `${pkgs.lxd-image-server}/bin/lxd-image-server init` doesn't actually need to run as `root` as long as all required paths exist and are owned by the appropriate user - is that correct?

If you use `LogsDirectory` and `systemd.tmpfiles.rules` then I believe all file creation is handled and you don't need to run as `root` at all.

If I'm incorrect then you want `ExecStartPre = "+/path/to/pre-start-script";`. Take a look at **Table 1. Special executable prefixes** under https://www.freedesktop.org/software/systemd/man/systemd.service.html#ExecStart= for details.

Reply:
I've seen that it's just cumbersome, it'd rather just do regular preStart and then have some magic in between add the +

Perhaps that should be added.

Reply:
Running an `ExecStartPre` as `root` should be cumbersome because generally there is no need to do it. \</opinion>

So the question is... can we get away without running as `root` at all? I think we can...

Reply:
https://github.com/NixOS/nixpkgs/pull/144444 now it is

Reply:
> tmpfiles "d" directive wants to cleanup the directory aswell... not sure if that's what i want

`tempfiles.d` does not want to cleanup a directory unless you pass it a cleanup argument.


Running a script as `root` is seldom needed. When there was a convenient option (`PermissionsStartOnly = true;`) almost **every single** NixOS module used it, even when there was absolutely no need to. Forcing developers to think and learn the proper way to do things seems like a good idea to me in exchange for 1 in 100 modules needing an inconvenient use of `ExecStartPre = "+" + (pkgs.writeScript "init.sh" "code to run as root")`. Additionally, `ExecStartPre` takes a list so you can combine `root` and non `root` scripts. Note that in your implementation you bunched everything together to run as `root`. See how convenient that was?

Reply:
hrm, makes sense